builtin/logical/pki/crl_util.go

```diff
@@ -0,0 +1,153 @@
+package pki
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+type revocationInfo struct {
+	CertificateBytes []byte `json:"certificate_bytes"`
+	RevocationTime   int64  `json:"unix_time"`
+}
+
+func revokeCert(req *logical.Request, serial string) (*logical.Response, error) {
+	certEntry, userErr, intErr := fetchCertBySerial(req, "revoked/", serial)
+	if certEntry != nil {
+		return nil, nil
+	}
+
+	certEntry, userErr, intErr = fetchCertBySerial(req, "certs/", serial)
+	switch {
+	case userErr != nil:
+		return logical.ErrorResponse(userErr.Error()), nil
+	case intErr != nil:
+		return nil, intErr
+	}
+
+	// Possible TODO: use some kind of transaction log in case of an
+	// error anywhere along here (so we've validated that we got a
+	// value back, but want to make sure it not only is deleted from
+	// certs/ but also shows up in revoked/ and a CRL is generated)
+	err := req.Storage.Delete("certs/" + serial)
+
+	if err != nil {
+		return nil, fmt.Errorf("Error deleting cert from valid-certs location")
+	}
+
+	cert, err := x509.ParseCertificate(certEntry.Value)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing certificate")
+	}
+	if cert == nil {
+		return nil, fmt.Errorf("Got a nil certificate")
+	}
+
+	if cert.NotAfter.Before(time.Now()) {
+		return nil, nil
+	}
+
+	revInfo := revocationInfo{
+		CertificateBytes: certEntry.Value,
+		RevocationTime:   time.Now().Unix(),
+	}
+
+	certEntry, err = logical.StorageEntryJSON("revoked/"+serial, revInfo)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating revocation entry")
+	}
+
+	err = req.Storage.Put(certEntry)
+	if err != nil {
+		return nil, fmt.Errorf("Error saving revoked certificate to new location")
+	}
+
+	err = buildCRL(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error encountered during CRL building: %s", err)
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"revocation_time": revInfo.RevocationTime,
+		},
+	}, nil
+}
+
+func buildCRL(req *logical.Request) error {
```

Comment:
I think we should maybe put a lock in this at the top, just to prevent any concurrent CRL construction since they always write to the same key in the storage backend.